### ISSUE TITLE
How can we access to the dataset, source codes and pre-trained models?

### ISSUE THREAD

**Issue by surefyyq** [YimuWang96/YimuWang96.github.io#1]
I'm surprised that after 8 months there's still nothing here.
May i ask when and how can i get the dataset and source codes?

Thanks

**Comment by YimuWang96** [comment]
Seriously? Have you checked the page of our project?

**Comment by surefyyq** [comment]
appreciate your reply, 
but forgive me for being stupid, I just can't find the source code in the project page, or is it the kind of project that cannot be open sourced because of the company's participation?
i'm so impressed by the SA module in your paper which capturing fine-grained parts, the visualization seems nice.
Is it possible to read your source code?
respect

**Comment by YimuWang96** [comment]
> appreciate your reply,
> but forgive me for being stupid, I just can't find the source code in the project page, or is it the kind of project that cannot be open sourced because of the company's participation?
> i'm so impressed by the SA module in your paper which capturing fine-grained parts, the visualization seems nice.
> Is it possible to read your source code?
> respect

That's one of the reason and thus we changed the name of our dataset (it was called another name initially instead of DA Retail).
Another reason is that due to the COVID19 I can not reach the final code which is placed in school.

But I can give you a 'pseudo-code' of this part. I think that will not violate the contract.

batch is a batch of images (B x C x H x W) while featuremap is the corresponding feature map at the last convolutional layer.

`
    def cropBatch(self, batch, featuremap):
        featuremap = featuremap
        featuremap = np.sum(featuremap.data.cpu().numpy(), axis=1)

        newBatch = []

        for i in range(len(batch)):
            tmp_heat = featuremap[i]
            tmp_heat -= np.sum(tmp_heat) / 36
            tmp_heat = np.maximum(tmp_heat, 0)
            tmp_heat /= np.max(tmp_heat)
            tmp_heatmap = np.uint8(255 * tmp_heat)
            _, binary = cv2.threshold(tmp_heatmap, 127, 255, cv2.THRESH_BINARY)

            _1, contours, _2 = cv2.findContours(binary, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
            tmp_max = -1
            tmp_maxi = -1
            for i in range(len(contours)):
                cnt = contours[i]
                _, _, w, h = cv2.boundingRect(cnt)
                if w * h > tmp_max:
                    tmp_max = w * h
                    tmp_maxi = i
            tmpx, tmpy, tmpw, tmph = cv2.boundingRect(contours[tmp_maxi])
            tmpx1, tmpy1, tmpx2, tmpy2 = int(tmpx * 227 / 6), int(tmpy * 227 / 6), int(
                math.ceil((tmpx + tmpw) * 227 / 6)), int(math.ceil((tmpy + tmph) * 227 / 6))
            tmp_img = batch[i].data.cpu().numpy()
            tmp_img = tmp_img.transpose(1, 2, 0)
            tmp_img = Image.fromarray(np.uint8(tmp_img))
            tmp_bbox = (tmpx1, tmpy1, tmpx2, tmpw)
            tmp_bbox = tuple(tmp_bbox)
            tmp_img = tmp_img.crop(tmp_bbox).resize((227, 227))
            tmpiimg = np.asarray(tmp_img)

            newBatch.append(tmpiimg)

        newBatch = np.array(newBatch).transpose(0, 3, 1, 2)
        return self.to_var(torch.from_numpy(newBatch).float())

`

Best Regards

**Comment by surefyyq** [comment]
Thanks for your answer! I think I get your idea about SA module
